bindings/sundials4py/idas/idas.cpp

```diff
@@ -0,0 +1,319 @@
+        });
+
+  m.def("IDASensInit",
+        [](void* ida_mem, int Ns, int ism, std::function<IDASensResStdFn> resS,
+           std::vector<N_Vector> yS0, std::vector<N_Vector> ypS0)
+        {
+          auto fn_table  = get_idas_fn_table(ida_mem);
+          fn_table->resS = nb::cast(resS);
+          return IDASensInit(ida_mem, Ns, ism, idas_resS_wrapper, yS0.data(),
+                             ypS0.data());
+        });
+
+  ///
+  // IDAS adjoint bindings
+  ///
+
+  m.def("IDAInitB",
+        [](void* ida_mem, int which,
+           std::function<std::remove_pointer_t<IDAResFnB>> resB,
+           sunrealtype tB0, N_Vector yyB0, N_Vector ypB0)
+        {
+          int ida_status = IDAInitB(ida_mem, which, idas_resB_wrapper, tB0,
+                                    yyB0, ypB0);
+
+          auto fn_table = idasa_user_supplied_fn_table_alloc();
+          auto idab_mem = static_cast<IDAMem>(IDAGetAdjIDABmem(ida_mem, which));
+          idab_mem->python = fn_table;
+
+          ida_status = IDASetUserDataB(ida_mem, which, idab_mem);
+          if (ida_status != IDA_SUCCESS)
+          {
+            free(fn_table);
+            throw sundials4py::error_returned(
+              "Failed to set user data in IDAS memory");
+          }
+
+          fn_table->resB = nb::cast(resB);
+          return ida_status;
+        });
+
+  m.def("IDAQuadInitB",
+        [](void* ida_mem, int which,
+           std::function<std::remove_pointer_t<IDAQuadRhsFnB>> resQB,
+           N_Vector yQBO)
+        {
+          auto fn_table   = get_idasa_fn_table(ida_mem, which);
+          fn_table->resQB = nb::cast(resQB);
+          return IDAQuadInitB(ida_mem, which, idas_resQB_wrapper, yQBO);
+        });
+
+  BIND_IDAB_CALLBACK(IDASetJacFnB, IDALsJacFnB, lsjacfnB, idas_lsjacfnB_wrapper,
+                     nb::arg("ida_mem"), nb::arg("which"),
+                     nb::arg("jacB").none());
+
+  BIND_IDAB_CALLBACK2(IDASetPreconditionerB, IDALsPrecSetupFnB, lsprecsetupfnB,
+                      idas_lsprecsetupfnB_wrapper, IDALsPrecSolveFnB,
+                      lsprecsolvefnB, idas_lsprecsolvefnB_wrapper,
+                      nb::arg("ida_mem"), nb::arg("which"),
+                      nb::arg("psetB").none(), nb::arg("psolveB").none());
+
+  BIND_IDAB_CALLBACK2(IDASetJacTimesB, IDALsJacTimesSetupFnB, lsjactimessetupfnB,
+                      idas_lsjactimessetupfnB_wrapper, IDALsJacTimesVecFnB,
+                      lsjactimesvecfnB, idas_lsjactimesvecfnB_wrapper,
+                      nb::arg("ida_mem"), nb::arg("which"),
+                      nb::arg("jsetupB").none(), nb::arg("jtimesB").none());
+
+  m.def("IDAInitBS",
+        [](void* ida_mem, int which, std::function<IDAResStdFnBS> resBS,
+           sunrealtype tB0, N_Vector yyB0, N_Vector ypB0)
+        {
+          int ida_status = IDAInitBS(ida_mem, which, ida_resBS_wrapper, tB0,
+                                     yyB0, ypB0);
```

Comment:
```suggestion
          if (!resBS) { throw sundials4py::illegal_value("resBS was null"); }
          int ida_status = IDAInitBS(ida_mem, which, ida_resBS_wrapper, tB0,
                                     yyB0, ypB0);
          if (ida_status != IDA_SUCCESS)
          {
            throw sundials4py::error_returned("Failed to initialize IDA");
          }
```
Should we return the error flag?